File: scripts/generate_og_image.py
from pathlib import Path

from PIL import Image, ImageDraw, ImageFilter, ImageFont


ROOT = Path(__file__).resolve().parents[1]
SOURCE_IMAGE = ROOT / "public" / "campaign" / "tombra-portrait.jpg"
OUTPUT_IMAGE = ROOT / "public" / "og-image.png"

CANVAS_SIZE = (2400, 1260)
FINAL_SIZE = (1200, 630)


def load_font(size: int, bold: bool = True) -> ImageFont.FreeTypeFont:
    names = (
        "Arial Bold.ttf",
        "Arial.ttf",
        "DejaVuSans-Bold.ttf",
        "DejaVuSans.ttf",
    )
    roots = (
        Path("/System/Library/Fonts/Supplemental"),
        Path("/Library/Fonts"),
        Path("/usr/share/fonts/truetype/dejavu"),
    )

    for root in roots:
        for name in names if bold else reversed(names):
            path = root / name
            if path.exists():
                return ImageFont.truetype(str(path), size)

    return ImageFont.load_default()


def draw_shadowed_text(
    canvas: Image.Image,
    position: tuple[int, int],
    text: str,
    font: ImageFont.FreeTypeFont,
    fill: str,
) -> None:
    shadow_layer = Image.new("RGBA", canvas.size, (0, 0, 0, 0))
    shadow_draw = ImageDraw.Draw(shadow_layer)
    x, y = position
    shadow_draw.text((x + 5, y + 5), text, font=font, fill=(0, 0, 0, 105))
    shadow_layer = shadow_layer.filter(ImageFilter.GaussianBlur(2))
    canvas.paste(shadow_layer.convert("RGB"), (0, 0), shadow_layer)
    ImageDraw.Draw(canvas).text(position, text, font=font, fill=fill)


def main() -> None:
    portrait = Image.open(SOURCE_IMAGE).convert("RGB")
    width, height = CANVAS_SIZE
    canvas = Image.new("RGB", CANVAS_SIZE, "#063f2a")
    draw = ImageDraw.Draw(canvas)

    draw.rectangle([0, 0, width, height], fill="#063f2a")
    draw.polygon([(0, 0), (960, 0), (780, height), (0, height)], fill="#083820")
    draw.polygon([(880, 0), (1220, 0), (1060, height), (720, height)], fill="#0a5a39")
    draw.polygon([(0, 980), (1160, 850), (width, 1000), (width, height), (0, height)], fill="#052f22")
    draw.rectangle([0, 0, 30, height], fill="#f7ce4f")

    frame_x, frame_y = 1240, 220
    frame_w, frame_h = 1120, 788
    shadow = Image.new("RGBA", (frame_w + 80, frame_h + 80), (0, 0, 0, 0))
    shadow_draw = ImageDraw.Draw(shadow)
    shadow_draw.rounded_rectangle([28, 28, frame_w + 52, frame_h + 52], radius=42, fill=(0, 0, 0, 95))
    shadow = shadow.filter(ImageFilter.GaussianBlur(24))
    canvas.paste(shadow.convert("RGB"), (frame_x - 40, frame_y - 40), shadow)

    portrait_large = portrait.resize((frame_w, frame_h), Image.Resampling.LANCZOS)
    mask = Image.new("L", (frame_w, frame_h), 0)
    mask_draw = ImageDraw.Draw(mask)
    mask_draw.rounded_rectangle([0, 0, frame_w, frame_h], radius=42, fill=255)
    canvas.paste(portrait_large, (frame_x, frame_y), mask)
    draw.rounded_rectangle(
        [frame_x - 8, frame_y - 8, frame_x + frame_w + 8, frame_y + frame_h + 8],
        radius=50,
        outline="#f7ce4f",
        width=10,
    )

    draw.rounded_rectangle([136, 150, 492, 168], radius=9, fill="#f7ce4f")
    draw_shadowed_text(canvas, (136, 245), "TOMBRA 2027", load_font(66), "#f7ce4f")
    draw_shadowed_text(canvas, (136, 420), "Tombra", load_font(168), "#ffffff")
    draw_shadowed_text(canvas, (136, 585), "Mohammed", load_font(168), "#ffffff")
    draw_shadowed_text(canvas, (144, 785), "A Stronger Voice for", load_font(72), "#ffffff")
    draw_shadowed_text(canvas, (144, 900), "Bayelsa West", load_font(102), "#f7ce4f")

    badge_x, badge_y, badge_w, badge_h = 136, 1068, 820, 108
    draw.rounded_rectangle([badge_x, badge_y, badge_x + badge_w, badge_y + badge_h], radius=54, fill="#ffffff")
    badge_text = f"Jobs {chr(8226)} Representation {chr(8226)} Grassroots"
    draw.text(
        (badge_x + 78, badge_y + 30),
        badge_text,
        font=load_font(42),
        fill="#07543a",
    )

    final = canvas.resize(FINAL_SIZE, Image.Resampling.LANCZOS)
    final.save(OUTPUT_IMAGE, optimize=True)

    print(f"Source: {SOURCE_IMAGE.relative_to(ROOT)} ({portrait.width}x{portrait.height})")
    print(f"Output: {OUTPUT_IMAGE.relative_to(ROOT)} ({FINAL_SIZE[0]}x{FINAL_SIZE[1]})")
    print(f"Size: {OUTPUT_IMAGE.stat().st_size} bytes")


if __name__ == "__main__":
    main()

File: scripts/generate-og-image.mjs
import fs from "node:fs/promises";
import path from "node:path";
import sharp from "sharp";

const root = process.cwd();
const sourcePath = path.join(root, "public", "campaign", "tombra-portrait-high-res.jpg");
const outputPath = path.join(root, "public", "og-image.png");
const draftSize = { width: 2400, height: 1260 };
const finalSize = { width: 1200, height: 630 };

const escapeHtml = (value) =>
  value
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");

const portraitBuffer = await fs.readFile(sourcePath);
const portraitMetadata = await sharp(portraitBuffer).metadata();
const portraitDataUrl = `data:image/jpeg;base64,${portraitBuffer.toString("base64")}`;

const svg = `
<svg width="${draftSize.width}" height="${draftSize.height}" viewBox="0 0 ${draftSize.width} ${draftSize.height}" xmlns="http://www.w3.org/2000/svg">
  <defs>
    <filter id="textShadow" x="-10%" y="-10%" width="120%" height="120%">
      <feDropShadow dx="8" dy="8" stdDeviation="3" flood-color="#000000" flood-opacity="0.35"/>
    </filter>
  </defs>

  <rect width="2400" height="1260" fill="#063f2a"/>
  <image href="${portraitDataUrl}" x="780" y="0" width="1620" height="1260" preserveAspectRatio="xMidYMid slice"/>
  <rect x="0" y="0" width="1128" height="1260" fill="#063f2a"/>
  <polygon points="0,0 930,0 745,1260 0,1260" fill="#083820" opacity="0.8"/>
  <polygon points="845,0 1215,0 1025,1260 665,1260" fill="#0b5a38" opacity="0.72"/>
  <rect x="935" y="0" width="190" height="1260" fill="#063f2a" opacity="0.45"/>
  <rect x="1045" y="0" width="120" height="1260" fill="#0b5a38" opacity="0.38"/>
  <rect x="1100" y="0" width="132" height="1260" fill="#0b4d38" opacity="0.42"/>
  <rect x="1198" y="0" width="118" height="1260" fill="#3f6d60" opacity="0.26"/>
  <rect x="1284" y="0" width="72" height="1260" fill="#0b5a38" opacity="0.12"/>
  <polygon points="0,985 1150,850 1190,1260 0,1260" fill="#052f22" opacity="0.86"/>

  <rect x="136" y="150" width="356" height="18" rx="9" fill="#f7ce4f"/>
  <g filter="url(#textShadow)" font-family="Arial, Helvetica, sans-serif" font-weight="800">
    <text x="136" y="305" font-size="66" fill="#f7ce4f" letter-spacing="2">TOMBRA 2027</text>
    <text x="136" y="570" font-size="168" fill="#ffffff">Tombra</text>
    <text x="136" y="735" font-size="168" fill="#ffffff">Mohammed</text>
    <text x="144" y="850" font-size="72" fill="#ffffff">A Stronger Voice for</text>
    <text x="144" y="995" font-size="102" fill="#f7ce4f">Bayelsa West</text>
  </g>

  <rect x="136" y="1068" width="820" height="108" rx="54" fill="#ffffff"/>
  <text x="214" y="1138" font-family="Arial, Helvetica, sans-serif" font-size="42" font-weight="800" fill="#07543a">Jobs • Representation • Grassroots</text>
</svg>`;

const draftBuffer = await sharp(Buffer.from(svg)).png().toBuffer();

await sharp(draftBuffer)
  .resize(finalSize.width, finalSize.height, {
    kernel: sharp.kernel.lanczos3,
  })
  .png({
    compressionLevel: 6,
    adaptiveFiltering: true,
  })
  .toFile(outputPath);

const outputMetadata = await sharp(outputPath).metadata();
const outputStats = await fs.stat(outputPath);

console.log(
  `Source: ${escapeHtml(path.relative(root, sourcePath))} (${portraitMetadata.width}x${portraitMetadata.height})`,
);
console.log(`Draft: ${draftSize.width}x${draftSize.height}`);
console.log(`Output: ${escapeHtml(path.relative(root, outputPath))} (${outputMetadata.width}x${outputMetadata.height})`);
console.log(`Size: ${outputStats.size.toLocaleString()} bytes`);
